### PR TITLE
[PUBDEV-7706] Unignore ignored columns that are used

### PR DESCRIPTION
**Pull request by honzasterba**
- improves flow UX
- makes our REST API easier to use
- is backwards compatible, only eliminates error states

### CODE REVIEW

File: h2o-core/src/main/java/hex/ModelBuilder.java [hunk 1]
```diff
@@ -1243,8 +1243,16 @@ public void init(boolean expensive) {
...
     if( _parms._ignored_columns != null ) {
-      _train.remove(_parms._ignored_columns);
-      if( expensive ) Log.info("Dropping ignored columns: "+Arrays.toString(_parms._ignored_columns));
+      Set<String> ignoreColumnSet = new HashSet<>(Arrays.asList(_parms._ignored_columns));
+      if (_parms._response_column != null) ignoreColumnSet.remove(_parms._response_column);
+      if (_parms._fold_column != null) ignoreColumnSet.remove(_parms._fold_column);
+      if (_parms._offset_column != null) ignoreColumnSet.remove(_parms._offset_column);
+      if (_parms._weights_column != null) ignoreColumnSet.remove(_parms._weights_column);
+      Collection<String> otherUsedColumns = getAdditionalUsedColumns();
+      for (String otherUsedColumn : otherUsedColumns) ignoreColumnSet.remove(otherUsedColumn);
```
Comment (satai) [inline h2o-core/src/main/java/hex/ModelBuilder.java:1252]: removeAll

**Comment by michalkurka** [comment]
I don't like this approach personally but I understand why it is done.

The main issue I have with this is that this approach is really a hit-and-miss. People need to remember to implement the getAdditionalUsedColumns and they need to do it correctly (which is hard to test IMHO).

What will happen when I have a column used eg. in GLM interactions and ignored in feature columns?

File: h2o-core/src/main/java/hex/ModelBuilder.java (hunk 1 above)
Comment (honzasterba) [reply h2o-core/src/main/java/hex/ModelBuilder.java:1252]: fixed

**Comment by honzasterba** [comment]
> I don't like this approach personally but I understand why it is done.

this basically duplicates what is done in Py and R clients, I dont like the code duplication as well, but its better solution than replicate the logic one more time in Flow

> The main issue I have with this is that this approach is really a hit-and-miss. People need to remember to implement the getAdditionalUsedColumns and they need to do it correctly (which is hard to test IMHO).

well not exactly, if you dont, only Flow will be a little harder to use, which is easy to fix

> What will happen when I have a column used eg. in GLM interactions and ignored in feature columns?

I am not sure if interactions are the same case as coxph(start,stop) columns, but if they are user will just get an error that they are ignoring a column that is necessary

File: h2o-py/h2o/estimators/estimator_base.py [hunk 2]
```diff
@@ -526,13 +526,6 @@ def _requires_training_frame(self):
         """
         return True
 
-    def _additional_used_columns(self, parms):
```
Comment (honzasterba) [inline h2o-py/h2o/estimators/estimator_base.py:529]: simmilar cleanup could be done in R client

**Comment by koniecsveta** [comment]
LGTM - will you clean up also R client in this PR?

**Comment by honzasterba** [comment]
> LGTM - will you clean up also R client in this PR?

probably yes, if this approach is approved by @michalkurka

**Review by wendycwong — APPROVED** [review]
LGTM.  Hope Michalk will turn around too.

**Review by michalkurka — APPROVED** [review]
Thank you!

**Pull request merged**